src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -321,138 +384,104 @@ int MPL_gavl_tree_free(MPL_gavl_tree_t gavl_tree)
-            if (gavl_tree_iptr->gavl_free_fn)
-                gavl_tree_iptr->gavl_free_fn((void *) dnode->val);
+            if (tree_ptr->gavl_free_fn)
+                tree_ptr->gavl_free_fn((void *) dnode->val);
             MPL_free(dnode);
         }
     }
-    MPL_free(gavl_tree_iptr);
+    MPL_free(tree_ptr);
     return mpl_err;
 }
 
-int MPL_gavl_tree_delete(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len)
+static void gavl_tree_delete_node(gavl_tree_s * tree_ptr, gavl_tree_node_s * dnode, void **val)
 {
-    int mpl_err = MPL_SUCCESS;
-    gavl_tree_node_s *cur_node;
     gavl_tree_node_s *inorder_node;
-    gavl_tree_s *gavl_tree_iptr = (gavl_tree_s *) gavl_tree;
 
-    DECLARE_STACK(gavl_tree_node_s *, node_stack);
+    if (dnode->right == NULL) {
+        /* no right child, next inorder node is parent */
+        if (dnode->parent == NULL) {
+            /* delete root node */
+            if (dnode->left) {
+                tree_ptr->root = dnode->left;
+                tree_ptr->root->parent = NULL;
+            } else {
+                tree_ptr->root = NULL;
+            }
+            *val = (void *) dnode->val;
+            MPL_free(dnode);
+        } else {
+            /* assign deleted node's left child to its parent */
+            inorder_node = dnode->parent;
+            if (inorder_node->left == dnode)
+                inorder_node->left = dnode->left;
+            else
+                inorder_node->right = dnode->left;
+            if (dnode->left)
+                dnode->left->parent = inorder_node;
+
+            *val = (void *) dnode->val;
+            MPL_free(dnode);
+            TREE_STACK_PUSH(tree_ptr, inorder_node);
+        }
+    } else {
+        /* find next inorder node and replace deleted node with it */
+        inorder_node = dnode->right;
+        TREE_STACK_PUSH(tree_ptr, dnode);
+        while (inorder_node->left) {
+            TREE_STACK_PUSH(tree_ptr, inorder_node);
+            inorder_node = inorder_node->left;
+        }
+
+        if (inorder_node->parent != dnode) {
+            if (inorder_node->right)
+                inorder_node->right->parent = inorder_node->parent;
+            inorder_node->parent->left = inorder_node->right;
+        } else {
+            /* only right child of deleted node */
+            dnode->right = NULL;
+        }
+        *val = (void *) dnode->val;
+        dnode->addr = inorder_node->addr;
+        dnode->len = inorder_node->len;
+        dnode->val = inorder_node->val;
+        MPL_free(inorder_node);
+    }
+
+    if (TREE_STACK_EMPTY(tree_ptr)) {
+        tree_ptr->cur_node = NULL;
+    } else {
+        TREE_STACK_POP(tree_ptr, tree_ptr->cur_node);
+    }
+    return;
+}
+
+int MPL_gavl_tree_delete(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len)
+{
+    int mpl_err = MPL_SUCCESS;
+    gavl_tree_node_s *dnode;
+    gavl_tree_s *tree_ptr = (gavl_tree_s *) gavl_tree;
 
     void *val;
     do {
-        CLEAR_STACK(node_stack);
```

Comment:
Can the `do{if (tree_ptr->root == NULL) exit; ... } while(1)` loop be simplified to `while(tree_ptr->root){...}`?